Add user settings module

lukpueh opened the pull request:
The newly added module provides methods to read environment variables and rcfiles and to write them to the `in_toto.settings` module, i.e. override them (fix #137).

The PR also adds a call of the settings parsing and overriding method to in-toto's `__init__.py`, which should be the first thing that's executed when using in_toto. 
Any code that accesses `in_toto.settings` thereafter receives the values defined by the user.

More information about how the user sets the settings can be found in the [header docstring of `user_settings.py`](https://github.com/in-toto/in-toto/blob/4390aeae19fd9e8cee69790acd76510298c812ec/in_toto/user_settings.py#L15-L48).

Commit:
Add module to read envvars and rcfiles

This module provides methods to read environment variables and
rcfiles and to write them to the in_toto.settings module, i.e.
override them.

The commit also adds a call of the settings parsing and overriding
method to in-toto's __init__.py, which should be the first thing
that's executed when using in_toto. Any code that accesses
in_toto.settings thereafter receives the values defined by the
user.

Commit:
Add unit tests for user settings module

vladimir-v-diaz commented on in_toto/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from in_toto.user_settings import set_settings
```
Our code style guidelines advises against importing items from a module: https://github.com/secure-systems-lab/code-style-guidelines#imports

What's wrong with the following?

``` Python
import in_toto.user_settings

in_toto.user_settings_setting.set_settings()
```

vladimir-v-diaz commented on in_toto/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from in_toto.user_settings import set_settings
+# Override settings in settings.py
```
Can you put some blank lines between import statements and the body of the module?

vladimir-v-diaz commented on in_toto/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from in_toto.user_settings import set_settings
+# Override settings in settings.py
+set_settings()
```
It seems wrong to perform actions when a project or module is imported.

Also, are you careful to catch exceptions or errors when your project is imported?  It would be strange to get an error about a missing or corrupted config file when attempting to import the project.

lukpueh replied:
Good point, I actually had a similar feeling.  It's probably better to only monkey patch the settings module when using a command line tool. What do you think?

vladimir-v-diaz replied:
Yes, it makes more sense to me to have the command-line tool read the config file(s) and environment variables.  The command tool help message or its documentation should explain how the settings can be modified via the environment variables and config files.  It should also list the supported settings.

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+  them to the in_toto.settings module (override them).
+
+  Notes:
+  - Variable values are converted to lists if they contain colons
```
Should this be "environment values", so as not to confuse it with a Python variable?

lukpueh replied:
Values form envvars or rcfiles. I will rephrase.

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+
+  Notes:
+  - Variable values are converted to lists if they contain colons
+  - Variable names are converted to upper case
```
To be precise, do you mean that in-toto settings are always upper case?

Can you be a bit more precise here, perhaps providing an example?  Are `in-toto.settings.SOME_SETTING` settings always upper case regardless of the way that environments variables are written (i.e., upper case, lower case, or mixed case) ?

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+  them to the in_toto.settings module (override them).
+
+  Notes:
+  - Variable values are converted to lists if they contain colons
```
In a comment, you should explain why this convention is used.   It is arbitrarily chosen, is it to ease parsing, to avoid parsing issues, or because multiple values are set this way with environment variables?

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+    in_toto.settings.ARTIFACT_EXCLUDES = ["*.link", ".gitignore"]
+
+  - RCfile values follow the format expected by Python's builtin
+    ConfigParser. The format requires at least one section, sections however
```
Do you need to revise this comment?  "The format requires at least one section, sections however are ignored"

Section headers are ignored?  Except that one section header is required?

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+USER = os.path.expanduser("~")
+
+RC_PATHS = [
+  os.path.join("/etc", "in_toto", "config"),
```
Can you list these in the order of precedence used in the docstring?  It helps readability, and a reviewer to verify that an RC path isn't missing.

lukpueh replied:
I guess I rather reverse the list in the docstring. If I reversed here, I'd have to do a `RC_PATHS.reverse()` below where I use the list, or iterate backwards, which seems odd too.

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+  os.path.join(USER, ".config", "in_toto"),
+  os.path.join(USER, ".in_toto", "config"),
+  os.path.join(USER, ".in_totorc"),
+  ".in_totorc"
```
Why is this ".in_totorc" written on a separate line, unlike the others?

lukpueh replied:
Each line is an item in the list. `.in_totorc` just looks different because it is a relative path and therefore does not need to be `os.path.join`ed.

vladimir-v-diaz replied:
Ahh, true.

Although it's overkill, I wonder if `os.path.join(os.getcwd(), ".in_totorc")` is more readable.

lukpueh replied:
I think that's an overkill. :)

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+ENV_PREFIX = "IN_TOTO_"
+USER = os.path.expanduser("~")
+
+RC_PATHS = [
```
How come this list doesn't include the `.in_totorc` RC file in the current working directory?  See line 40.

lukpueh replied:
This list does include `.in_totorc`. See your [comment below](https://github.com/in-toto/in-toto/pull/143#discussion_r148293274).

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+
+
+def get_rc():
+  """Load RCfiles from the usual places and return a dictionary of
```
The usual places?  Can you be more explicit here?  How about, "the paths specified in `RC_PATHS`?

Preferably, you should assume that a developer/reader is only looking at this docstring and is unaware of the module's other functions and docstrings.

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+  return env_dict
+
+
+def get_rc():
```
How come this function's docstring isn't formatted according to our code style guidelines?  Where is the section on exceptions, return values, purpose, etc.? 

The other functions in this module are also missing proper docstrings.

lukpueh replied:
I thought that we only use this convention for API functions. But you are right, it says
*"for every function [..] that will be called from other modules"*.

I'll change this.

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+  user_settings.update(get_rc())
+
+  for name, value in user_settings.iteritems():
+    setattr(SETTINGS_MODULE, name, value)
```
Is it possible to populate `in-toto.settings` with unexpected variables or values?  For example, `in-toto.settings.[]`  How do you expect the module to catch these exceptions or fail in these instances?

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+  See LICENSE for licensing information.
+
+<Purpose>
+  Test user settings module.
```
I find the following comment easier to determine which module is being tested:

> Test in-toto.user_settings.py.

vladimir-v-diaz reviewed (commented):
My comments for the main modules, I still need to review `test_user_settings.py`

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+  return rc_dict
+
+
+def set_settings():
```
Is it wise to first backup the current settings before attempting to overwrite them?

lukpueh replied:
Not unless we use the backup for something (e.g. rollback on error). But considering that we use a [whitelist](https://github.com/in-toto/in-toto/pull/143#discussion_r148387751), which reduces the unexpected error likelihood; and only use this function in [cli tools](https://github.com/in-toto/in-toto/pull/143#discussion_r148377987), where I'd rather exit on error, I think we don't need to backup.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+
+    os.environ["IN_TOTO_ARTIFACT_EXCLUDES"] = "e:n:v"
+    os.environ["IN_TOTO_artifact_basepath"] = "e/n/v"
+    os.environ["IN_TOTO_NEW_ENV_SETTING"] = "new env setting"
```
Is this manually setting some environment variables?  Is it possible to temporarily set some environment variables and verify that environment variables are properly loaded?

Perhaps testing actual environment variables on the system is unnecessary, but something to consider.

lukpueh replied:
I thought about it, but that would actually be a test for `os.environ`.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+import shutil
+import tempfile
+import in_toto.settings
+from in_toto.user_settings import set_settings, get_env, get_rc
```
I guess it's okay to import individual items from a module (in this unit test), but you should get into the habit of using absolute import statements.

```
in-toto.user_settings.get_rc()
```

```
set_settings()
```

The former is more readable in my opinion, since it shows where the function comes from.  The latter can be confused for some function in this module.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+    for key in dir(in_toto.settings):
+      self.settings_backup[key] = getattr(in_toto.settings, key)
+
+    # Create an rc file and set some environment variables
```
Another thing you can do is to have a pre-generated `.in_totorc` that you can load in the unit tests.  This approach might be faster, easier for others to inspect and review, etc.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+
+  @classmethod
+  def tearDownClass(self):
+    os.chdir(self.working_dir)
```
If you use pre-generated config files, another thing it free you from is the need to change directories and clean up temp files.  The unit tests would only need to load config files.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+    self.assertEquals(rc_dict["NEW_RC_SETTING"], "new rc setting")
+
+
+  def test_get_env(self):
```
Should you test what happens if an empty environment is loaded?  That is, there are no in-toto environment variables set, and no in-toto config files can be loaded?

lukpueh replied:
Why not. Will add.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+          "ARTIFACT_EXCLUDES=r:c:file\n"
+          "new_rc_setting=new rc setting")
+
+    os.environ["IN_TOTO_ARTIFACT_EXCLUDES"] = "e:n:v"
```
Can you test a value that contains special characters like `$`, `%`, `..`, etc.?  What other interesting things?  Python identifiers?  Function calls?

lukpueh replied:
I don't evaluate the strings. So it should be fine.

vladimir-v-diaz replied:
Is is possible for someone to attempt a `import in-toto.settings` and get errors because of unexpected entries?

lukpueh replied:
I added sanitization for the available settings where used in 4072cb3.

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+  user_settings.update(get_rc())
+
+  for name, value in user_settings.iteritems():
+    setattr(SETTINGS_MODULE, name, value)
```
Can you verify that someone can't cause unintended lines?  Is it possible to populate `in-toto.settings.py` with malicious input like: `"OS.REMOVE(my_files)".lower()`

Perhaps you don't have to worry about malicious input in config files and environment variables, but something to ponder.

lukpueh replied:
It does not seem possible, since I am assigning module variables with strings.

However, there isn't really a need to allow everything, so it might be better to make a whitelist, i.e. only allow settings for which there is a default in `settings.py`.

vladimir-v-diaz replied:
Using a whitelist is the more secure approach in my opinion.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+    self.assertFalse("NOT_IN_TOTO_NOTHING" in env_dict)
+
+
+  def test_set_settings(self):
```
Is it possible to test the import behavior that follows from the modified `__init__()`?  These test conditions appear to test environment and RCfiles indirectly.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -0,0 +1,96 @@
+    self.assertEquals(in_toto.settings.NEW_RC_SETTING, "new rc setting")
+
+    # Settings that are only in env are still used even if they are new
+    self.assertEquals(in_toto.settings.NEW_ENV_SETTING, "new env setting")
```
Should you have a test condition that verifies overwritten settings?  Suppose `in-toto.settings.MY_SETTING` is already set, and you attempt to call `set_settings()`.  Is `MY_SETTING` overwritten, as expected?

lukpueh replied:
`ARTIFACT_BASEPATH` and `ARTIFACT_EXCLUDES` have defaults in `settings.py`. So the overriding is already tested.

vladimir-v-diaz replied:
I see.  Should the test and/or modules indicate that `ARTIFACT_BASEPATH` and `ARTIFACT_EXCLUDES` are actual in-toto settings?  If you happen to remove these settings in the future, do these tests conditions become invalid (in terms of failing to test for the overwriting behavior)?  It's good to note assumptions in the comments for these sort of situations

Also, assuming that `user_settings.py` overwrites settings in `in-toto.settings.py`, wouldn't a more accurate test (1) verify that the new setting is an actual supported setting in `in-toto.settings.py` (2) note the value of the setting and (3) verify that the value in (2) is overwritten with the new value?

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -0,0 +1,132 @@
+"""
+import os
+import ConfigParser
+import in_toto.settings as SETTINGS_MODULE
```
Is there a reason why you use `import X as FOO`?  In this case,
`SETTINGS_MODULE` is only one character longer, and less readable than `in_toto.settings`.

vladimir-v-diaz reviewed (commented):
Okay, here are my remaining comments for this pull request.

Commit:
Replace `from` imports related to user settings

As pointed out by @vladimir-v-diaz in #143, our style guidelines
advice against importing indiviudal items from modules.

Commit:
Rephrase purpose of `test_user_settings.py`

Commit:
Use pre-generated rcfile for user_settings tests

This commit minimizes the test setup/teardown by adding a pre-
generated rcfile instead of writing a temp dir and an rcfile
on the fly, which should make the test slightly faster and maybe
more readable.

Commit:
Restrict user_settings to whitelisted settings

There is no need for unlimited monkeypatching of the settings
module using envvars and rcfiles, since in-toto only uses a known
set of settings anyway.
Moreover, a whitelist approach seems more secure.

This commit also updates the relevant unittest, renames a variable
and extends some comments and docstrings.

Commit:
Move test rcfile to separate directory

This commit moves the test `.in_totorc` file from the `demo_files`
dir to a new `rc_test` dir.
Other unit tests copy all files from `demo_files` to their cwd, and
would therefor load the settings from `.in_totorc` although they
rely on the defaults in `settings.py`.

This commit also removes the default settings backup and restore
in test_user_settings.py's setup/teardown methods, because they are
not needed.

Commit:
Move user settings from __init__ to cli tools

As pointed out by @vladimir-v-diaz it is odd to perform actions
like monkeypatching a module using env variables and rcfiles
when importing a package or module.

This behavior is better suited when the package is used via
command line tools. Hence, this commit removes the a call to
`user_settings.set_settings` from __init__ and instead calls the
function in the command line interfaces.

Commit:
Add logger to user_settings module

Call log.info if a setting is overridden

Commit:
Remove `LOG_LEVEL` from user setting whitelist

User settings (envvars and rcfiles) only allow strings, but the
setting `LOG_LEVEL` expects a constant from the Python `logging`
module.
Since `LOG_LEVEL` is likely to be removed (in-toto/in-toto#117)
we just ignore it for now and remove it from the user settings
whitelist.

Commit:
Update header and comments in `settings.py`

Commit:
Add section about (user) settings to README

Commit:
Remove user settings to uppercase

The case of settings defined in envvars or rcfiles is no longer
ignored, that is the currently whitelisted settings must be defined
as such:
ARTIFACT_BASE_PATH and ARTIFACT_EXCLUDES in rcfiles
and IN_TOTO_ARTIFACT_BASE_PATH and IN_TOTO_ARTIFACT_EXCLUDES
in environment variables.

Since the user settings module monkey patches (whitelisted)
settings it feels more secure to be more restrictive, e.g. by
also requiring the case to match.

Commit:
Update the user settings header and function docs

Following @vladimir-v-diaz's advice this commit updates the
docstrings in the user_settings module.
Details can be found in the function docstrings, which conform
with our style guidelines.
The header gives only an overview and references the respective
functions.

lukpueh commented:
@vladimir-v-diaz, thanks for you comments. I tried to address them accordingly. Let me know what you think.

Update: It seems that  2ea0e5d broke it, because `ConfigParser` converts names to lower case. I'll fix this on Monday. You can review the rest if you want though.

Commit:
Enable case-sensitivity for rcfile options

Adds "reset" of `optionxform` default conversion to enable
case sensitivity in rcfile optiosn parsed with Python's
`ConfigParser`.

https://docs.python.org/2/library/configparser.html#ConfigParser.RawConfigParser.optionxform

Commit:
Add/update settings sanitizing in runlib

Both of the currently available settings are used in
`runlib.record_artifacts_as_dict`.
This commits adds a new Exception subclass
`in_toto.exceptions.SettingsError` and updates the runlib function
to raise this error if any of the two settings is invalid.

Commit:
Update user_settings tests + doc typo fixes

Updates user_settings unittests following @vladimir-v-diaz's
comments in in-toto/in-toto#143 and fixes typos in user_settings
docstrings.

lukpueh commented:
I fixed the remaining issues regarding case sensitivity that broke the unittests last week, added some testcases and fixed missing settings sanitization where they are used. Mind giving it another pass, @vladimir-v-diaz?

vladimir-v-diaz commented on README.md:
```diff
@@ -105,6 +105,16 @@ in-toto-verify --layout <layout path>
+these settings can be configured using environment variables or RCfiles.
+Take a look at [`in_toto.settings`](https://github.com/in-toto/in-toto/blob/develop/in_toto/settings.py)
+to learn more about the available defaults and at
+[`in_toto.user_settings`](https://github.com/in-toto/in-toto/blob/develop/in_toto/user_settings.py)
```
I get a 404 error with this link.

vladimir-v-diaz replied:
I'm not so sure about the  documentation of this section.  Is the command-line tool intended to be run by a variety of users?  If so, wouldn't linking to a Python module might be inappropriate for these users?

lukpueh replied:
Re 404: should have used relative path (`user_settings.py` isn't in the main branch yet).
Re inappropriate: Agreed, will change.

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -119,7 +120,7 @@ def record_artifacts_as_dict(artifacts):
 
   <Exceptions>
-    None.
+    SettingsError if ARTIFACT_BASE_PATH or ARTIFACT_EXCLUDES can't be used
```
It might be better if this listed the exception name: `in-toto.exceptions.SettingsError`, since the caller will have to specify the full name to correctly catch the exception.

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -137,18 +138,29 @@ def record_artifacts_as_dict(artifacts):
       os.chdir(in_toto.settings.ARTIFACT_BASE_PATH)
-    except OSError as e:
-      raise OSError("Review your ARTIFACT_BASE_PATH setting - {}".format(e))
+    except Exception as e:
```
Is there a reason why you switched from catching `OSError` to catching `Exception`?  According to the [Python documentation](https://docs.python.org/3/library/os.html#module-os), all functions in the os modules raise `OSError` 

> in the case of invalid or inaccessible file names and paths, or other arguments that have the correct type, but are not accepted by the operating system.

lukpueh replied:
It could also be a TypeError. I figured I just catch all Exceptions and re-raise as `SettingsError`.

vladimir-v-diaz replied:
You can also explicitly catch all the exceptions that you're aware of.  For the unexpected exceptions, it might be better to let the exception pass through. 
```
try:
    os.chdir(...)
except (OSError, TypeError) as e:
    ...
```

A problem I have encountered in the past is that the code would sometimes catch weird things like `NameError` and the exception message would be totally wrong.  You append the original exception, so I guess it would be easy to debug / catch these cases.

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -137,18 +138,29 @@ def record_artifacts_as_dict(artifacts):
   for path in artifacts:
     norm_artifacts.append(os.path.normpath(path))
 
+  # If ARTIFACT_EXCLUDES is set it must be a list of strings or an empty
```
Or an empty what?

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -137,18 +138,29 @@ def record_artifacts_as_dict(artifacts):
-      raise OSError("Review your ARTIFACT_BASE_PATH setting - {}".format(e))
+    except Exception as e:
+      raise in_toto.exceptions.SettingsError(
+          "Review your ARTIFACT_BASE_PATH setting '{}' - {}".format(
```
Consider providing more information on what could have gone wrong.  For example, an invalid path.  On the other hand, appending the exception, as you have done, would make it clear what is wrong with `ARTIFACT_BASE_PATH`.

lukpueh replied:
I think adding the exception message is good enough. I guess we can always enhance this later, e.g. when taking care of #117.

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -137,18 +138,29 @@ def record_artifacts_as_dict(artifacts):
+  # If ARTIFACT_EXCLUDES is set it must be a list of strings or an empty
+  # TODO: Change NAMES_SCHEMA to something more semantically accurate
+  if (in_toto.settings.ARTIFACT_EXCLUDES and not
+      securesystemslib.formats.NAMES_SCHEMA.matches(
```
Maybe PATH_SCHEMA or PATHS_SCHEMA is a bit more accurate?  I am not certain, but ARTIFACT_EXCLUDES appears to be glob patterns?

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -12,59 +12,47 @@
-  os.path.join(USER, ".config", "in_toto"),
-  os.path.join(USER, ".in_toto", "config"),
-  os.path.join(USER, ".in_totorc"),
+  os.path.join(USER_PATH,  ".config", "in_toto", "config"),
```
I guess home directory would be another good name for this.

vladimir-v-diaz commented on in_toto/settings.py:
```diff
@@ -12,36 +12,35 @@
 # in-toto-run/in-toto-record
-# FIXME: This is likely to become a command line argument
 # FIXME: Do we want different base paths for materials and products?
 ARTIFACT_BASE_PATH = None
```
Should ARTIFACT_BASE_PATH be set to the user's home directory, by default?  Or current working directory?

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -76,18 +64,47 @@ def _colon_split(value):
+    converted into a list.
+
+
+    Example:
```
👍

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -76,18 +64,47 @@ def _colon_split(value):
+    None.
+
+  <Side Effects>
+    Calls function to read files from disk.
```
For a while I used to list similar entries in Side Effects, but others have pointed out that this isn't really a side effect.

lukpueh replied:
I remember having this discussion and also that there were uncertainties about some things and whether they would qualify as side effect. But I am pretty sure that there was no doubt about I/O.

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -98,35 +115,93 @@ def get_env():
+    None.
+
+  <Side Effects>
+    Calls function to read files from disk.
```
Another instance where someone might argue that this isn't really a side effect.

vladimir-v-diaz commented on in_toto/user_settings.py:
```diff
@@ -98,35 +115,93 @@ def get_env():
+    if user_setting:
+      log.info("Using setting {0}={1}".format(
+          setting, user_setting))
+      setattr(in_toto.settings, setting, user_setting)
```
Consider adding an else-statement that logs a debug message for settings that are ignored/invalid.

lukpueh replied:
Adding that else-statement would actually log the default settings and not the ignored/invalid settings, which might be more interesting anyway. I will add this.

vladimir-v-diaz commented on test/test_user_settings.py:
```diff
@@ -34,63 +32,72 @@ class TestUserSettings(unittest.TestCase):
-          "artifact_basepath=r/c/file\n"
-          "ARTIFACT_EXCLUDES=r:c:file\n"
-          "new_rc_setting=new rc setting")
+    # We use `demo_files` as test dir because it has an `.in_totorc`, which
```
I don't understand this comment.  Is `demo_files` supposed to be `rc_test`?

lukpueh replied:
Oh, I forgot to update this comment. I first added the rcfile to demo_files, as you suggested, but then discovered that this would break other tests that copied everything from demo_files and therefor also loaded and used the test rcfile.

vladimir-v-diaz reviewed (commented):
Here is the requested review of your follow-on commits.  I had minor comments.  Feel free to merge when you feel the pull request is ready.

Commit:
Enhance settings section in README.md

Commit:
Misc user settings comments/log fixes

Addressing @vladimir-v-diaz's review regarding comments and
log statements related to user settings.

Commit:
Move user settings patching after args parsing

Since parsing the user settings has verbose output, it is
better to postpone it until after parsing command line tool
arguments, based on which e.g. the verbosity level of the program
is set.

Commit:
Refactor ARTIFACT_EXCLUDES

The setting lists patterns to exclude artifacts, hence the new
better suited name: ARTIFACT_EXCLUDE_PATTERNS

vladimir-v-diaz commented on README.md:
```diff
@@ -122,7 +122,7 @@ Also, setting values that contain colons are parsed as list.
 ##### Available Settings
 
-`ARTIFACT_EXCLUDES` Specifies a list of glob patterns that can be used to
+`ARTIFACT_EXCLUDE_PATTERNS` Specifies a list of glob patterns that can be used to
```
:1st_place_medal:

vladimir-v-diaz commented:
LBTM (Looks Brilliant To Me)

:)
